### ISSUE TITLE
Portable regex-based syntax highlighting

### ISSUE THREAD

**Issue by dougalm** [google-research/dex-lang#176]
A bit of syntax highlighting goes a long way towards making things more readable. Right now we have a [little emacs mode](https://github.com/google-research/dex-lang/blob/main/misc/dex.el) and an independent [syntax highlighter for the HTML renderings](https://github.com/google-research/dex-lang/blob/main/src/lib/RenderHtml.hs#L67-L105) using parser combinators. Does anyone know of a portable solution that lets us define a bunch of regexes and generate modes for vim, emacs, sublime etc?

**Comment by paulyoung** [comment]
The closest I’ve come across is [Iro](https://eeyo.io/iro/documentation/) but that doesn’t support Vim or Emacs.

**Comment by hexahedria** [comment]
Perhaps one could daisy-chain that with something like [tm2deftheme](https://github.com/emacsfodder/tmtheme-to-emacs) or [coloration](https://github.com/sickill/coloration)?

**Comment by sureyeaah** [comment]
As already suggested by @paulyoung and @danieldjohnson, we can use Iro to cover (VSCode, Eclipse, Intellij, Sublime Code, Atom and a few others), daisy-chain that with tm2deftheme (for emacs) and coloration (for vim and a few others). I'll take this one :)

**Comment by sureyeaah** [comment]
![image](https://user-images.githubusercontent.com/2617705/101732910-f78afb80-3ae3-11eb-9dc3-787112579043.png)
This is what I came up with. I limited it to only check on a token basis instead of trying to match the actual parser so that it's flexible as the language's syntax might change. If this looks alright, I'll generate something for vim/emacs as well.

**Comment by apaszke** [comment]
That looks great! One minor point would be that we only use upper-case letters for types as a convention, not because it's a requirement (unlike in Haskell), so matching on that can sometimes incorrectly color regular functions such as `IToF`. But it's no big deal and this is already a huge step forward!

**Comment by sureyeaah** [comment]
@apaszke Gotcha, I'll publish this so we have something to start with and then have a follow-up PR with more context sensitive highlighting. 
- Should I add the plugin to this repo or a separate one?
- Is there something licensing related that I need to keep in mind before pubishing to vscode marketplace?

**Comment by joaogui1** [comment]
So, what's the state of this issue? It looks pretty useful @sureyeaah @apaszke

**Comment by sureyeaah** [comment]
@joaogui1 Forgot about this, will send a PR this week.

**Comment by apaszke** [comment]
@sureyeaah adding stuff to this repo sounds reasonable for now. I don't know anything about the licensing issues of vscode marketplace and I suspect that it depends on what you want to put up there. Self-contained scripts that define syntax highlighters that are entirely authored by you should be ok. Distributing any Haskell components would be more tricky, because we would have to bundle all the licenses of our dependencies.

**Comment by joaogui1** [comment]
Pinging @sureyeaah

**Comment by sureyeaah** [comment]
@joaogui1 sorry, I'm working on this right now.

**Comment by sharadmv** [comment]
For what it's worth, I wrote a simple vim plugin: https://github.com/sharadmv/dex-vim.

Dex code looks like this:
![162616510_1253415871720566_9017564163822846057_n](https://user-images.githubusercontent.com/1142926/111859727-2f35ea80-8900-11eb-8cac-4f976a179074.png)

Hope it's helpful!

**Comment by sureyeaah** [comment]
@sharadmv Thanks! I'll merge this PR and then if it's alright, I'll make some improvements to these plugins and your vim one as well.

**Comment by sharadmv** [comment]
Yes please! It's my first vim plugin so I don't entirely know what I'm doing.

**Comment by general-rishkin** [comment]
> ![image](https://user-images.githubusercontent.com/2617705/101734020-b1cf3280-3ae5-11eb-9fa4-cf2820a8035d.png)
> 
> This is what I came up with for VS Code. I limited it to only check on a token basis instead of trying to mimic the entire parser so that it's flexible as the language's syntax might change. If this looks alright, I'll publish the plugin and try to recreate this for vim.

Is this still active. How can I get a hold of it for use in VSCode?

**Comment by sureyeaah** [comment]
Need someone to take ownership of this PR :(

**Comment by axch** [comment]
For the record, PR #503 was an attempt here; reading the conversation there, it was close, but didn't quite make it over the line.  Could be picked up to complete this issue, or one could start fresh.

**Comment by benjamin-macadam** [comment]
Would a tree-sitter grammar make sense? This would be a two birds/one stone situation for me, since I'd also like to get a better handle on how Tree-sitter works.

**Comment by apaszke** [comment]
Yeah, tree-sitter seems really promising! That would be great.

**Comment by youqad** [comment]
Oops, I just saw this! For what it’s worth, as I couldn’t find a syntax highlighting extension for Dex in the Marketplace, I wrote a little VS Code extension based the regexes of the Emacs Lisp file: 
https://marketplace.visualstudio.com/items?itemName=youqad.dex-syntax-highlighting

**Comment by richardwesthaver** [comment]
any luck with a tree-sitter-dex implementation? I would be interested in working on that but not sure how it aligns with some of the current project goals (https://github.com/google-research/dex-lang/issues/1304) - may not be a priority for the project if `dex web` is the IDE and the LSP to support it is WIP. I would love to add dex to my list of tree-sitter parsers though :)
